Compact module index

not-my-profile opened the pull request:
| Before | After |
|--|--|
| ![image](https://user-images.githubusercontent.com/73739153/152498211-5f55fdc4-a16e-4263-9788-12a26d1f353d.png) | ![image](https://user-images.githubusercontent.com/73739153/152634448-6ad58820-3fdc-4a95-917f-995cb4ea1043.png)

Closes #465.

tristanlatr commented on pydoctor/templatewriter/summary.py:
```diff
@@ -26,13 +26,26 @@ def moduleSummary(module: model.Module, page_url: str) -> Tag:
+        pass
+        div = tags.div(class_='submodules')
+        for m in sorted(contents, key=fullName):
+            div(tags.code(tags.a(m.name, href=m.url)))
```
Please use epydoc2stan.taglink to create links to objects in the system.

not-my-profile replied:
Why? Of the three `if`s in epydoc2stan.taglink none will ever apply:

* `if not o.isVisible`: not the case because we filter modules for visibility before
* `if label is None:` we would need to specify the label since we don't want the full name
* `if url.startswith(page_url + '#'):` also will never apply

So using it seems quite pointless to me.

tristanlatr replied:
Because we want to centralize the link generation at only one place such that we can add custom css for internal vs external links without too much hazard. See #468

tristanlatr commented on pydoctor/templatewriter/summary.py:
```diff
@@ -26,13 +26,26 @@ def moduleSummary(module: model.Module, page_url: str) -> Tag:
+    if len(contents) > 50 and not any(any(s.submodules()) for s in contents):
+        # If there are more than 50 modules and no submodule has
+        # further submodules we use a more compact presentation.
+        pass
```
```suggestion
```

tristanlatr reviewed (changes requested):
I’m not sure that we absolutely need css customization for this corner case.
Couldn’t we do with a simple list item? 

Also, I know it’s a pain, but it would be good to have a test for that new corner case. 

About the model refractor, I think it’s good to have this new method. I’de like to read @adiroiban opinion on this.

tristanlatr commented on pydoctor/templatewriter/summary.py:
```diff
@@ -26,13 +26,26 @@ def moduleSummary(module: model.Module, page_url: str) -> Tag:
     def fullName(obj: model.Documentable) -> str:
         return obj.fullName()
+
+    if len(contents) > 50 and not any(any(s.submodules()) for s in contents):
```
What if there is 50 sub modules and one package? 
I think the 50 du modules should still be grouped together while the packages would be extended.

Also packages are instances of model.Package, so maybe it could work with an `isinstance` check instead ?

not-my-profile replied:
I believe that having more than 50 submodules and one package is really uncommon. I don't think I have ever seen that in a Python package, have you? If not I'd rather keep it simple.

tristanlatr replied:
Very well

tristanlatr commented on pydoctor/templatewriter/summary.py:
```diff
@@ -26,13 +26,26 @@ def moduleSummary(module: model.Module, page_url: str) -> Tag:
+        # If there are more than 50 modules and no submodule has
+        # further submodules we use a more compact presentation.
+        pass
+        div = tags.div(class_='submodules')
```
Maybe we could read a text like: « 50 submodules: <the list>»

not-my-profile replied:
I am afraid I don't know what you mean?

tristanlatr replied:
I'm suggesting to show a text saying how many modules are listed under a single list item, such that reader have a better view of the tree. Also cannot misunderstand the entry as being only one module.

not-my-profile replied:
Now that the submodules are displayed in a list item the vertical bar is displayed which IMO really helps with indicating that these are modules. So I don't think adding a label is necessary.

not-my-profile commented:
> I’m not sure that we absolutely need css customization for this corner case.
Couldn’t we do with a simple list item?

I changed it to a list item. Though I still have two lines of CSS:

* `list-style: none` to disable the list bullet (because one bullet on the module index represents one module, using one bullet for multiple modules would give the wrong impression)
* `line-height: 1.8em` because otherwise the text is very vertically bunched together which makes it hard to read

> Also, I know it’s a pain, but it would be good to have a test for that new corner case.

I added a test case.

tristanlatr commented on pydoctor/themes/base/apidocs.css:
```diff
@@ -114,6 +114,11 @@ li {
     padding-bottom: 5px;
 }
 
+.submodules {
```
```suggestion
/*special css class used only in the modules index when there is more than 50 modules on the same package*/
.submodules {
```

not-my-profile replied:
I updated the CSS selector to `#summaryTree .compact-modules` which is more descriptive. Is that descriptive enough or should I still add a comment? I certainly think we should avoid mentioning the exact number in the CSS, it's an implementation detail that is subject to change.

tristanlatr reviewed (approved):
Looks good for me! Thanks

not-my-profile commented on pydoctor/templatewriter/summary.py:
```diff
@@ -26,15 +26,31 @@ def moduleSummary(module: model.Module, page_url: str) -> Tag:
+            span(tags.code(epydoc2stan.taglink(m, m.url, label=m.name)))
+            span(', ')
+            if m.isPrivate:
+                span(class_='private')
```
I updated the PR to add the `private` class for private modules, which is necessary because otherwise they're always shown.

tristanlatr replied:
Thanks, I did not see that bug. It looks like the whole span will be hidden though. Where only one tag link and the commas should be private I think.

not-my-profile replied:
Yes it it's intentional that the whole span is hidden because otherwise there would be ,,,,,, left over.

tristanlatr replied:
Sorry it’s my fault I though is was one span for the whole thing.

tristanlatr commented on pydoctor/templatewriter/summary.py:
```diff
@@ -26,15 +26,32 @@ def moduleSummary(module: model.Module, page_url: str) -> Tag:
+        li.children[-1].children.pop() # type: ignore
+        ul(li)
+    else:
+        for m in sorted(contents, key=fullName):
```
```suggestion
        for m in sorted(contents, key=objects_order):
```

tristanlatr commented on pydoctor/test/test_templatewriter.py:
```diff
@@ -469,3 +469,28 @@ def func():
+    assert len(ul.children) == 50   # type: ignore
+
+    # the 51th module triggers the compact summary
+    fromText('', parent_name='top', modname='yet_another_sub', system=system)
+
+    ul = moduleSummary(top, '').children[-1]
+    assert ul.tagName == 'ul'       # type: ignore
+    assert len(ul.children) == 1    # type: ignore
```
```suggestion
    fromText('', parent_name='top', modname='_yet_another_sub', system=system)

    ul = moduleSummary(top, '').children[-1]
    assert ul.tagName == 'ul'       # type: ignore
    assert len(ul.children) == 1    # type: ignore
    
    # test that the last module is private
    assert 'private' in ul.children[0].children[-1].attributes['class'] # type: ignore
```

tristanlatr commented on pydoctor/templatewriter/summary.py:
```diff
@@ -26,15 +26,32 @@ def moduleSummary(module: model.Module, page_url: str) -> Tag:
+            span(tags.code(epydoc2stan.taglink(m, m.url, label=m.name)))
+            span(', ')
+            if m.isPrivate:
+                span(class_='private')
```
Added a suggestion in the test case to cover this branch.

adiroiban commented:
@tristanlatr if you guys are happy with the changes, feel free to merge merge it.

I have very little interest into pydoctor and generic API doc tools.
I always prefer to read the source code and most api docs is part of the source code.

I know that API doc tools are important.

I am helping with this project just because Twisted is using it and I care about Twisted... and there was no one else reviewing the PR...so I tried to help and unblock the project.

Many thanks to both of you for you help with pydoctor!

tristanlatr commented:
@not-my-profile are you willing to address the changes? I don't really care about using `objects_order` but a test case for the private modules would be good. 

Thanks

tristanlatr commented on README.rst:
```diff
@@ -108,6 +108,7 @@ pydoctor 22.2.0
   this allows to have the external intershinx links blue while the internal links are red.
 * Smarter line wrapping in summary and parameters tables.
 * Any code inside of ``if __name__ == '__main__'`` is now excluded from the documentation.
+* The module index now uses a more compact presentation for modules with more than 50 submodules and no subsubmodules.
```
I'll move this to the right place.
